### PR TITLE
3DS2 Transaction version from API

### PR DESCRIPTION
**Pull request by caiofaustino**
## Summary
- Use API response do define 3DS2 Transaction version.
- Add missing parameter to Action and Token JSON modules.
- Refactor all 3DS2 classes to Kotlin. 🎉 
- Use Coroutines instead of ThreadPoolExecutor.

### CODE REVIEW

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+            }
+            Threeds2Action.ACTION_TYPE -> {
+                val threeds2Action = action as Threeds2Action
+                if (threeds2Action.token.isNullOrEmpty()) {
+                    throw ComponentException("3DS2 token not found.")
+                }
+                if (threeds2Action.subtype == null) {
+                    throw ComponentException("3DS2 Action subtype not found.")
+                }
+                val subtype = parse(threeds2Action.subtype.orEmpty())
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt:123]: Maybe it's clearer to use `SubType.parse` here and remove the `SubType.Companion.parse` import?

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        closeTransaction(getApplication())
+    }
+
+    override fun timedout() {
+        Logger.d(TAG, "challenge timed out")
+        notifyException(Authentication3DS2Exception("Challenge timed out."))
+        closeTransaction(getApplication())
+    }
+
+    override fun protocolError(protocolErrorEvent: ProtocolErrorEvent) {
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt:159]: Minor, you can do something like this:

```kotlin
    override fun protocolError(protocolErrorEvent: ProtocolErrorEvent) {
        with(protocolErrorEvent.errorMessage) {
            Logger.e(TAG, "protocolError - $errorCode - $errorDescription - $errorDetails")
            notifyException(Authentication3DS2Exception("Protocol Error - $this"))
        }
        closeTransaction(getApplication())
    }
```

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+            notifyException(CheckoutException("Error starting challenge", e))
+        }
+    }
+
+    @Throws(ComponentException::class)
+    fun createEncodedFingerprint(authenticationRequestParameters: AuthenticationRequestParameters): String {
+        val fingerprintJson = JSONObject()
+        try {
+            // TODO getMessageVersion is not used?
+            fingerprintJson.put("sdkAppID", authenticationRequestParameters.sdkAppID)
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt:258]: Minor: you can use `fingerprintJson.apply` or `with(authenticationRequestParameters)` here

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+            fingerprintJson.put("sdkTransID", authenticationRequestParameters.sdkTransactionID)
+        } catch (e: JSONException) {
+            throw ComponentException("Failed to create encoded fingerprint", e)
+        }
+        return Base64Encoder.encode(fingerprintJson.toString())
+    }
+
+    private fun createChallengeParameters(challenge: ChallengeToken): ChallengeParameters {
+        // TODO referenceNumber, URL and version are not used?
+        val challengeParameters = ChallengeParameters()
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt:271]: Minor, can use `return ChallengeParameters().apply` here

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        } catch (e: JSONException) {
+            throw ComponentException("Failed to create challenge details", e)
+        }
+        return challengeDetails
+    }
+
+    companion object {
+        val TAG = LogUtil.getTag()
+
+        val PROVIDER: ActionComponentProvider<Adyen3DS2Component> = ActionComponentProviderImpl(
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt:315]: Does this need `@JvmField`? Same question for the other companion variables?

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/ChallengeToken.kt [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        private const val ACS_SIGNED_CONTENT = "acsSignedContent"
+        private const val ACS_TRANS_ID = "acsTransID"
+        private const val ACS_URL = "acsURL"
+        private const val MESSAGE_VERSION = "messageVersion"
+        private const val THREEDS_SERVER_TRANS_ID = "threeDSServerTransID"
+
+        @JvmField
+        val CREATOR = Creator(ChallengeToken::class.java)
+
+        val SERIALIZER: Serializer<ChallengeToken> = object : Serializer<ChallengeToken> {
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/ChallengeToken.kt:42]: Does this need `@JvmField` or `@JvmStatic`?

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/ChallengeToken.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            override fun serialize(modelObject: ChallengeToken): JSONObject {
+                val jsonObject = JSONObject()
+                try {
+                    jsonObject.putOpt(ACS_REFERENCE_NUMBER, modelObject.acsReferenceNumber)
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/ChallengeToken.kt:46]: You can also use `return JSONObject().apply`

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/FingerprintToken.kt [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+        @JvmField
+        val CREATOR: Parcelable.Creator<FingerprintToken> = Creator(FingerprintToken::class.java)
+
+        @JvmField
+        val SERIALIZER: Serializer<FingerprintToken> = object : Serializer<FingerprintToken> {
+            override fun serialize(modelObject: FingerprintToken): JSONObject {
+                val jsonObject = JSONObject()
+                try {
+                    jsonObject.putOpt(DIRECTORY_SERVER_ID, modelObject.directoryServerId)
```
Comment (jreij) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/FingerprintToken.kt:44]: You can also use `return JSONObject().apply`

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/ChallengeToken.kt (hunk 6 above)
Comment (caiofaustino) [reply 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/ChallengeToken.kt:42]: Good catch

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt (hunk 5 above)
Comment (caiofaustino) [reply 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt:315]: Good catch

**Review by jreij — APPROVED** [review]

**Pull request merged**